compiler/hash-ir/src/constant.rs

```diff
@@ -0,0 +1,537 @@
+                ConstKind::Scalar(Scalar::from_uint(value, size))
+            }
+        };
+
+        Const { ty, kind }
+    }
+
+    /// Create a boolean constant.
+    pub fn bool(value: bool) -> Self {
+        Self::new(COMMON_IR_TYS.bool, ConstKind::Scalar(Scalar::from_bool(value)))
+    }
+
+    /// Create a character constant.
+    pub fn char(value: char) -> Self {
+        Self::new(COMMON_IR_TYS.char, ConstKind::Scalar(Scalar::from(value)))
+    }
+
+    /// Create a new [Const] which represents a `usize`.
+    pub fn usize<C: HasDataLayout>(value: u64, ctx: &C) -> Self {
+        let kind = ConstKind::Scalar(Scalar::from_usize(value, ctx));
+        Self::new(COMMON_IR_TYS.usize, kind)
+    }
+}
+
+macro_rules! const_from_ty_impl {
+    ($($ty:ident),*) => {
+        $(
+            impl From<$ty> for Const {
+                fn from(value: $ty) -> Self {
+                    Const::new(COMMON_IR_TYS.$ty, ConstKind::Scalar(Scalar::from(value)))
+                }
+            }
+        )*
+    };
+}
+
+const_from_ty_impl!(f32, f64, char);
+
+/// The kind of constants that can be represented within the Hash IR.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum ConstKind {
+    /// Exotic constant, that has no size, effectively a unit.
+    Zero,
+
+    /// A constant scalar value.
+    Scalar(Scalar),
+
+    /// A pair is only used to represent `str` constants, which is the interned
+    /// string allocation, and the length. It is equivalent to the `Abi::Pair`
+    /// representation.
+    Pair {
+        /// The data pointer of the slice. This is represented as just
+        /// being an
+        data: InternedStr,
+
+        /// The length of the slice.
+        len: Scalar,
+    },
+
+    /// An allocated constant value, this is used to represent everything
+    /// that is non-scalar, including slices, ADTs and arrays.
+    Alloc {
+        /// The offset into the given allocation that is being used.
+        offset: Size,
+
+        /// The allocation itself.
+        alloc: AllocId,
+    },
+}
+
+impl ConstKind {
+    /// Create a [ConstKind] from a TIR literal.
+    pub fn from_lit<C: HasDataLayout>(lit: Lit, ctx: &C) -> Self {
+        match lit {
+            Lit::Int(int) => ConstKind::Scalar(Scalar::from(int.value())),
+            Lit::Float(float) => ConstKind::Scalar(Scalar::from(float.value())),
+            Lit::Str(str) => {
+                let data = str.interned_value();
+                ConstKind::Pair { data, len: Scalar::from_usize(data.len() as u64, ctx) }
+            }
+            Lit::Char(ch) => ConstKind::Scalar(Scalar::from(ch.value())),
+        }
+    }
+}
+
+/// A scalar value. [Scalar]s are used to represent all integer, characters, and
+/// floating point values, as well as integers. The largest scalar value is
+/// 128bits, i.e. a `u128` or `i128`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[repr(packed)]
+pub struct Scalar {
```

Comment:
This can definitely be used as a variant of `Term`, and the rest of the integer types can be data types defined using it